### PR TITLE
Fix swift-system to work on Windows.

### PR DESCRIPTION
**Pull request by al45tair**
The swift-system tests were not passing, for various reasons, mostly to do with path related confusion.

As part of this, add a function to create a temporary directory, as that was one of the problems.

Also add Windows support for pipe() and ftruncate(), which was missing.

rdar://125087707

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please test Windows platform

**Comment by al45tair** [comment]
I guess we don't test Windows for swift-system. We should.

File: Sources/System/FilePath/FilePathParsing.swift [hunk 1]
```diff
@@ -89,6 +89,11 @@ extension SystemString {
       // `_prenormalizeWindowsRoots` and resume.
       readIdx = _prenormalizeWindowsRoots()
       writeIdx = readIdx
+
+      // Skip redundant separators
+      while readIdx < endIndex && isSeparator(self[readIdx]) {
+          self.formIndex(after: &readIdx)
+      }
```
Comment (compnerd) [inline Sources/System/FilePath/FilePathParsing.swift:96]: This is kinda tricky, the duplicates are material at the beginning as the separator makes parsing differences. Internal duplicates are fine to normalize.

File: Sources/System/FilePath/FilePathTemp.swift [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+#if os(Windows)
+import WinSDK
+
+internal func getTemporaryDirectory() throws -> FilePath {
+  return try withUnsafeTemporaryAllocation(of: CInterop.PlatformChar.self,
+                                           capacity: Int(MAX_PATH) + 1) {
+    buffer in
+
+    let count = GetTempPath2W(DWORD(buffer.count), buffer.baseAddress)
+    if count == 0 {
```
Comment (compnerd) [inline Sources/System/FilePath/FilePathTemp.swift:33]: I'd collapse this with a `guard`.

File: Sources/System/Internals/Mocking.swift [hunk 3]
```diff
@@ -110,8 +110,8 @@ private var contextualMockingEnabled: Bool {
 extension MockingDriver {
   internal static var enabled: Bool { mockingEnabled }
 
-  internal static var forceWindowsPaths: Bool {
-    currentMockingDriver?.forceWindowsSyntaxForPaths ?? false
+  internal static var forceWindowsPaths: Optional<Bool> {
```
Comment (compnerd) [inline Sources/System/Internals/Mocking.swift:113]: Why the different spelling here? `Optional<Bool>` vs `Bool?` above.

File: Sources/System/Internals/Mocking.swift [hunk 4]
```diff
@@ -128,7 +128,7 @@ internal var mockingEnabled: Bool {
 }
 
 @inline(__always)
-internal var forceWindowsPaths: Bool {
+internal var forceWindowsPaths: Optional<Bool> {
```
Comment (compnerd) [inline Sources/System/Internals/Mocking.swift:131]: Similar

File: Sources/System/Internals/Syscalls.swift [hunk 5]
```diff
@@ -118,20 +118,90 @@ internal func system_dup2(_ fd: Int32, _ fd2: Int32) -> Int32 {
...
 }
+
+internal func system_mkdir(
+    _ path: UnsafePointer<CInterop.PlatformChar>,
+    _ mode: CInterop.Mode
+) -> CInt {
+#if ENABLE_MOCKING
+  if mockingEnabled { return _mock(path: path, mode) }
+#endif
+  return mkdir(path, mode)
```
Comment (compnerd) [inline Sources/System/Internals/Syscalls.swift:142]: Is this shared? I think that we should prefer `CreateDirectoryW` over `mkdir`.

File: Sources/System/Internals/Syscalls.swift [hunk 6]
```diff
@@ -118,20 +118,90 @@ internal func system_dup2(_ fd: Int32, _ fd2: Int32) -> Int32 {
...
+}
+
+internal func system_rmdir(
+    _ path: UnsafePointer<CInterop.PlatformChar>
+) -> CInt {
+#if ENABLE_MOCKING
+  if mockingEnabled { return _mock(path: path) }
+#endif
+  return rmdir(path)
```
Comment (compnerd) [inline Sources/System/Internals/Syscalls.swift:151]: Similar, `RemoveDirectoryW` should be preferred.

File: Sources/System/Internals/WindowsSyscallAdapters.swift [hunk 7]
```diff
@@ -115,11 +114,133 @@ internal func pwrite(
...
+
+  return 0;
+}
+
+@inline(__always)
+internal func mkdir(
+  _ path: UnsafePointer<CInterop.PlatformChar>,
+  _ mode: CInterop.Mode
+) -> CInt {
+  return _wmkdir(path)
```
Comment (compnerd) [inline Sources/System/Internals/WindowsSyscallAdapters.swift:163]: I think we should prefer `CreateDirectoryW`. The win32 API is safer.

File: Sources/System/Internals/WindowsSyscallAdapters.swift [hunk 8]
```diff
@@ -115,11 +114,133 @@ internal func pwrite(
...
+}
+
+@inline(__always)
+internal func rmdir(
+  _ path: UnsafePointer<CInterop.PlatformChar>
+) -> CInt {
+  return _wrmdir(path)
```
Comment (compnerd) [inline Sources/System/Internals/WindowsSyscallAdapters.swift:170]: Similar, prefer the win32 api (`RemoveDirectoryW`).

File: Sources/System/FilePath/FilePathParsing.swift (hunk 1 above)
Comment (al45tair) [reply Sources/System/FilePath/FilePathParsing.swift:96]: Yes. This skips duplicates _after_ the beginning has been processed (that's what the line above is doing). The problem was that we weren't skipping duplicates immediately after what this code refers to as the "root" (i.e. after UNC paths or drive letters).

File: Sources/System/Internals/Syscalls.swift (hunk 5 above)
Comment (al45tair) [reply Sources/System/Internals/Syscalls.swift:142]: This is shared. We're actually using `_wmkdir` for Windows at present.

File: Sources/System/Internals/Syscalls.swift (hunk 6 above)
Comment (al45tair) [reply Sources/System/Internals/Syscalls.swift:151]: Somewhat asymmetrically, we're using `RemoveDirectoryW` on Windows (see `FilePathTemp.swift`), mostly because the code for `recursiveRemove` isn't shared whereas `makeDirectory` is. Though maybe `makeDirectory` should use `CreateDirectoryW` on Windows and set a security descriptor?

**Comment by al45tair** [comment]
@swift-ci Please test

File: Sources/System/FilePath/FilePathTemp.swift [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+      // If it was truncated, increase capaciy and try again
+      if len > buffer.count {
+        capacity = len
+        return nil
+      }
+      return FilePath(SystemString(platformString: buffer.baseAddress!))
+    }
+    if let path = path {
+      return path
+    }
```
Comment (lorentey) [inline Sources/System/FilePath/FilePathTemp.swift:135]: This seems _far_ too much logic for swift-system. This package is not supposed to be the place for building cross-platform abstractions.

File: Sources/System/FilePath/FilePathTemp.swift [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+ Copyright (c) 2024 Apple Inc. and the Swift System project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+*/
+
+// MARK: - API
+
+public func withTemporaryPath<R>(basename: FilePath.Component,
+                                 _ body: (FilePath) throws -> R) throws -> R {
```
Comment (lorentey) [inline Sources/System/FilePath/FilePathTemp.swift:13]: Nit: please follow the [stdlib's line breaking conventions](https://github.com/apple/swift/blob/main/docs/StandardLibraryProgrammersManual.md#line-breaking). 

```suggestion
public func withTemporaryPath<R>(
  basename: FilePath.Component,
  _ body: (FilePath) throws -> R
) throws -> R {
```

I cannot help questioning if this really needs to be a top-level function, rather than some static method on `FilePath`.

But more importantly: does this _really_ belong in swift-system?

File: Sources/System/FilePath/FilePathTemp.swift [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+}
+#endif
+
+internal let base64 = Array<UInt8>(
+  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_".utf8
+)
+
+internal func makeDirectory(at: FilePath) throws -> Bool {
+  return try at.withPlatformString {
+    if system_mkdir($0, 0o700) == 0 {
```
Comment (lorentey) [inline Sources/System/FilePath/FilePathTemp.swift:247]: If we want this function to create directories with the hardwired mode 700, then it should probably carry a more specific name than just `makeDirectory`.

File: Sources/System/FilePath/FilePathTemp.swift [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+internal func createUniqueTemporaryDirectory(
+  basename: FilePath.Component
+) throws -> FilePath {
+  var tempDir = try getTemporaryDirectory()
+  tempDir.append(basename)
+
+  while true {
+    tempDir.extension = createRandomString(length: 16)
```
Comment (lorentey) [inline Sources/System/FilePath/FilePathTemp.swift:275]: On POSIX systems, shouldn't we just call `mkdtemp` to make a unique temp directory?

**Review by lorentey — COMMENTED** [review]
Yay for the FilePath fixes! 🎉 Thanks 👍 

I have lots of grumpy comments about the additions. swift-system was never intended to be a POSIX emulation package, and I really dislike that it's being pushed to become that. (I don't think we have the means to do a good job of that, even if it was a good idea, which I don't think it is.) This package isn't for building cross-platform abstraction layers -- that's Foundation's job.

File: Sources/System/FilePath/FilePathTemp.swift [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+        return nil
+      }
+      return FilePath(SystemString(platformString: buffer.baseAddress!))
+    }
+    if let path = path {
+      return path
+    }
+  }
+  #else
+  return "/tmp"
```
Comment (lorentey) [inline Sources/System/FilePath/FilePathTemp.swift:138]: On POSIX systems, I think we should be looking at the `TMPDIR` envvar; we should not be hardwiring /tmp.

File: Sources/System/FilePath/FilePathTemp.swift (hunk 12 above)
Comment (al45tair) [reply Sources/System/FilePath/FilePathTemp.swift:278]: `mkdtemp` is an especially, egregiously horrible function. It writes to its argument, which sadly often works with string "constants" in C because they historically have often ended up in writeable memory. I agree that it's a standard function that does something very much like this. It's also awful.

File: Sources/System/FilePath/FilePathTemp.swift (hunk 13 above)
Comment (al45tair) [reply Sources/System/FilePath/FilePathTemp.swift:141]: Yes, that's a fair point, though I'm not certain getting the environment variables from `swift-system` is as easy as it should be. `getenv("TMPDIR")` in Swift doesn't appear to work for some reason I don't fully understand, and we don't want to be using Foundation from here.

File: Sources/System/Internals/WindowsSyscallAdapters.swift [hunk 14]
```diff
@@ -115,11 +114,144 @@ internal func pwrite(
...
-    // TODO(compnerd) map windows error to errno
+    ucrt._set_errno(mapWindowsErrorToErrno(GetLastError()))
     return Int(-1)
   }
   return Int(nNumberOfBytesWritten)
 }
 
+@inline(__always)
+internal func pipe(_ fds: UnsafeMutablePointer<Int32>) -> CInt {
+  return _pipe(fds, 4096, _O_BINARY | _O_NOINHERIT);
```
[the remark this answers is not in the document]
Comment (al45tair) [reply Sources/System/Internals/WindowsSyscallAdapters.swift:180]: Playing devil's advocate for a moment, the `_pipe()` API provided by the Windows C library is _very_ close to the `pipe()` API provided by POSIX. Yes, it has two extra arguments; the first is the buffer size (on POSIX, you get whatever the kernel decides, so hard-coding it on Windows is hardly a huge issue), and the second is some flags that let you set text mode or handle inheritance (neither of which POSIX really has). I wouldn't call this "emulation" so much as calling the equivalent function.

File: Sources/System/Internals/WindowsSyscallAdapters.swift [hunk 15]
```diff
@@ -115,11 +114,144 @@ internal func pwrite(
...
+}
+
+@inline(__always)
+internal func ftruncate(_ fd: Int32, _ length: off_t) -> Int32 {
```
[the remark this answers is not in the document]
Comment (al45tair) [reply Sources/System/Internals/WindowsSyscallAdapters.swift:184]: The entire C library on Windows is in some sense "emulated", since the C library is not the native platform API there. I think this likely explains the difference in approach from Saleem in the past. And yes, the UCRT doesn't have `ftruncate()` or any equivalent (that I'm aware of), but the Win32 API totally does all the things we need it to do to make that work, and it isn't terribly complicated to implement.

Note: I'm _not_ arguing for one moment that `swift-system` should become a POSIX emulation layer. Just observing that in practice, non-POSIX _largely_ means Windows at this point, and if anyone is using `FileDescriptor` for anything then they'll end up writing things that aren't portable to Windows unless we make sure that we have equivalent functionality. I don't think it makes sense to push yet more `#if os(Windows)` stuff up the stack.

(I'm also playing Devil's advocate here.)

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test

**Review by compnerd — APPROVED** [review]

File: Sources/System/FilePath/FilePathTempWindows.swift [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+/// - Parameters:
+///   - path: The directory to be deleted.
+///
+/// Removes a directory completely, including all of its contents.
+internal func _recursiveRemove(
+  at path: FilePath
+) throws {
+  // First, deal with subdirectories
+  try forEachFile(at: path) { findData in
+    if (findData.dwFileAttributes & DWORD(FILE_ATTRIBUTE_DIRECTORY)) != 0 {
```
[the remark this answers is not in the document]
Comment (al45tair) [reply Sources/System/FilePath/FilePathTempWindows.swift:76]: The file attribute constants are [documented to be a bit mask](https://learn.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants). I think the `!= 0` and `== 0` is idiomatic for Win32 code — not least because the alternative requires that the compiler optimise away an unnecessary comparison.

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test Linux platform

**Comment by al45tair** [comment]
@swift-ci Please test macOS platform

**Comment by al45tair** [comment]
@swift-ci Please test Windows platform

File: Sources/System/FilePath/FilePathTemp.swift [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+/// Create a temporary path for the duration of the closure.
+///
+/// - Parameters:
+///   - basename: The base name for the temporary path.
+///   - body: The closure to execute.
+///
+/// Creates a temporary directory with a name based on the given `basename`,
+/// executes `body`, passing in the path of the created directory, then
+/// deletes the directory and all of its contents before returning.
+public func withTemporaryFilePath<R>(
```
Comment (milseman) [inline Sources/System/FilePath/FilePathTemp.swift:21]: Make sure that any API additions go through some kind of review, ideally on the forums. At the very least, a dedicated PR that highlights it's an API addition PR as opposed to just a bug-fix PR for Windows.

For this function, would a static method on `FilePath` work instead of a global? E.g. `FilePath.withTemporaryPath { ... }`.

File: Sources/System/Internals/WindowsSyscallAdapters.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (lorentey) [reply Sources/System/Internals/WindowsSyscallAdapters.swift:192]: Anything importing SystemPackage is expected to end up having to deal with `#if os(Foo)` conditions, anyway. 

SystemPackage is not designed to provide cross-platform API.

[the remark this answers is not in the document]
Comment (al45tair) [reply Sources/System/Internals/WindowsSyscallAdapters.swift:192]: Sure. We spoke about this on Slack… fundamentally the issue here is that from a Windows developer's standpoint, a (kind of) subset of the POSIX file descriptor API does exist on Windows and is useful for porting tools written for UNIX. It isn't all that useful for Windows programming more generally — Windows developers would generally drop down to the Win32 layer rather than using `_open()` et al — everyone knows that it's all emulated anyway.

Basically, I don't think it's worth _not_ supporting things that are trivially easy to implement, like `pipe()` and `ftruncate()`. I also don't think it should evolve into a Cygwin like layer, because that's _way_ too far.

I thought we'd agreed that this was the right thing for Windows; maybe I misunderstood the conversation we had?

File: Sources/System/FilePath/FilePathTemp.swift (hunk 17 above)
Comment (al45tair) [reply Sources/System/FilePath/FilePathTemp.swift:21]: Yes, it would. My only reservation is that (to me) things like `FilePath.withTemporaryPath` read like a function that _constructs and returns_ a `FilePath` from some kind of `TemporaryPath` thing. This isn't the same as e.g. `Array`'s `withUnsafeBytes`, because that's called on an instance rather than on the type itself. Maybe that's just because I'm used to ObjC. It's obvious from the arguments that it takes a closure, of course.

I only needed this for the tests, so for now I think the thing to do is to make this non-public (I'll raise a PR to fix that). I'll check through to make sure I didn't leak any other new API and do the same there.
